File: mcmc/elixir/mcmc.exs
defmodule MetropolisHastings do
  # Target probability distribution: a mixture of two Gaussian distributions
  def target_distribution(x) do
    gaussian1 = 0.3 * :math.exp(-0.2 * :math.pow(x - 10, 2))
    gaussian2 = 0.7 * :math.exp(-0.2 * :math.pow(x + 10, 2))
    gaussian1 + gaussian2
  end

  # Metropolis-Hastings algorithm for MCMC sampling
  def metropolis_hastings(target, num_samples, initial_state, proposal_std) do
    current_state = initial_state
    accepted = 0
    samples = []

    for _ <- 1..num_samples do
      # Generate a proposal state from a normal distribution
      proposal = current_state + proposal_std * :rand.normal()

      # Calculate the acceptance probability
      acceptance_prob = min(1, target.(proposal) / target.(current_state))

      # Accept or reject the proposal
      if :rand.uniform() < acceptance_prob do
        current_state = proposal
        accepted = accepted + 1
      end

      # Save the current state as a sample
      samples = [current_state | samples]
    end

    acceptance_rate = accepted / num_samples
    IO.puts("Acceptance rate: #{acceptance_rate}")
    Enum.reverse(samples)
  end
end

# Define parameters for MCMC sampling
num_samples = 10000
initial_state = 0.0
proposal_std = 5.0

# Run MCMC sampling using Metropolis-Hastings algorithm
samples = MetropolisHastings.metropolis_hastings(&MetropolisHastings.target_distribution/1, num_samples, initial_state, proposal_std)

# Print the sampled distribution
Enum.each(samples, &IO.puts(&1))

File: mcmc/cs/mcmc.cs
using System;
using System.Collections.Generic;

namespace MCMC
{
    // Target probability distribution: a mixture of two Gaussian distributions
    class TargetDistribution
    {
        public static double Evaluate(double x)
        {
            double gaussian1 = 0.3 * Math.Exp(-0.2 * Math.Pow((x - 10), 2));
            double gaussian2 = 0.7 * Math.Exp(-0.2 * Math.Pow((x + 10), 2));
            return gaussian1 + gaussian2;
        }
    }

    // Metropolis-Hastings algorithm for MCMC sampling
    class MetropolisHastings
    {
        private Random random;

        public MetropolisHastings()
        {
            random = new Random();
        }

        public List<double> Run(double initial, double proposalStd, int numSamples)
        {
            List<double> samples = new List<double>();
            double currentState = initial;
            int accepted = 0;

            for (int i = 0; i < numSamples; i++)
            {
                // Generate a proposal state from a normal distribution
                double proposal = currentState + (proposalStd * random.NextDouble());

                // Calculate the acceptance probability
                double acceptanceProb = Math.Min(1, TargetDistribution.Evaluate(proposal) / TargetDistribution.Evaluate(currentState));

                // Accept or reject the proposal
                if (random.NextDouble() < acceptanceProb)
                {
                    currentState = proposal;
                    accepted++;
                }

                // Save the current state as a sample
                samples.Add(currentState);
            }

            double acceptanceRate = (double)accepted / numSamples;
            Console.WriteLine("Acceptance rate: " + acceptanceRate);
            return samples;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            // Define parameters for MCMC sampling
            int numSamples = 10000;
            double initial = 0;
            double proposalStd = 5;

            // Create an instance of the Metropolis-Hastings algorithm
            MetropolisHastings mcmc = new MetropolisHastings();

            // Run MCMC sampling using Metropolis-Hastings algorithm
            List<double> samples = mcmc.Run(initial, proposalStd, numSamples);

            // Print the sampled distribution
            foreach (var sample in samples)
            {
                Console.WriteLine(sample);
            }
        }
    }
}
